src/os/bluestore/BlueStore.cc

```diff
@@ -13727,6 +13734,94 @@ ObjectMap::ObjectMapIterator BlueStore::get_omap_iterator(
   return ObjectMap::ObjectMapIterator(new OmapIteratorImpl(logger,c, o, it));
 }
 
+int BlueStore::omap_iterate(
+  CollectionHandle &c_,   ///< [in] collection
+  const ghobject_t &oid, ///< [in] object
+  ObjectStore::omap_iter_seek_t start_from, ///< [in] where the iterator should point to at the beginning
+  std::function<omap_iter_ret_t(std::string_view, std::string_view)> f
+  )
+{
+  Collection *c = static_cast<Collection *>(c_.get());
+  dout(10) << __func__ << " " << c->get_cid() << " " << oid << dendl;
+  if (!c->exists) {
+    return -ENOENT;
+  }
+  std::shared_lock l(c->lock);
+  OnodeRef o = c->get_onode(oid, false);
+  if (!o || !o->exists) {
+    dout(10) << __func__ << " " << oid << "doesn't exist" <<dendl;
+    return -ENOENT;
+  }
+  o->flush();
+  dout(10) << __func__ << " has_omap = " << (int)o->onode.has_omap() <<dendl;
+  if (!o->onode.has_omap()) {
+    // nothing to do
+    return 0;
+  }
+
+  KeyValueDB::Iterator it;
+  {
+    auto bounds = KeyValueDB::IteratorBounds();
+    std::string lower_bound, upper_bound;
+    o->get_omap_key(string(), &lower_bound);
+    o->get_omap_tail(&upper_bound);
+    bounds.lower_bound = std::move(lower_bound);
+    bounds.upper_bound = std::move(upper_bound);
+    it = db->get_iterator(o->get_omap_prefix(), 0, std::move(bounds));
+  }
+
+  // seek the iterator
+  {
+    std::string key;
+    o->get_omap_key(start_from.seek_position, &key);
+    auto start = ceph::mono_clock::now();
+    if (start_from.seek_type == omap_iter_seek_t::LOWER_BOUND) {
+      it->lower_bound(key);
+      c->store->log_latency(
+        __func__,
+        l_bluestore_omap_lower_bound_lat,
+        ceph::mono_clock::now() - start,
+        c->store->cct->_conf->bluestore_log_omap_iterator_age);
+    } else {
+      it->upper_bound(key);
+      c->store->log_latency(
+        __func__,
+        l_bluestore_omap_upper_bound_lat,
+        ceph::mono_clock::now() - start,
+        c->store->cct->_conf->bluestore_log_omap_iterator_age);
+    }
+  }
+
+  // iterate!
+  std::string tail;
+  ceph::timespan next_lat_acc{0};
+  o->get_omap_tail(&tail);
+  while (it->valid()) {
+    std::string user_key;
+    if (const auto& db_key = it->raw_key().second; db_key >= tail) {
```

Comment:
`rocksdb::Slice::ToString()` sounds promising. Perhaps it's another place `std::string_view` could help to pacify.

```
                              - 9.46% _ZN14CFIteratorImpl7raw_keyB5cxx11Ev                                                                                                                                        ▒
                                 - 7.32% _ZNK7rocksdb5Slice8ToStringB5cxx11Eb                                                                                                                                     ▒
                                    - _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE10_M_replaceEmmPKcm                                                                                                    ▒
                                       - _ZNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEE9_M_mutateEmmPKcm                                                                                                   ▒
                                            2.69% _Znwm@plt                                                                                                                                                       ▒
```

Reply:
I think this is the leaf of our vast abstraction over the `rocksdb::Iterator`:

```
  string key() override {
    return dbiter->key().ToString();
  } 
  std::pair<std::string, std::string> raw_key() override { 
    return make_pair(prefix, key());
  } 
```

[`rocksdb::Slice` already offers `ToStringView()`](https://github.com/facebook/rocksdb/blob/main/include/rocksdb/slice.h#L95-L97) :-).